### PR TITLE
Tracking doc: rewrite `@wdio/utils` package into TypeScript

### PR DESCRIPTION
**Pull request by christian-bromann**
This is a tracking issue for organising the effort of re-writing the code base from vanilla JS into TypeScript. If you are reading this and are interested getting involved, this is a great chance to do so. Here are the rules:

- if you are interested converting source files from `/packages/wdio-utils` into TypeScript please comment on this thread and mention the files you want to convert to avoid duplicate work
- there is no reason to convert the whole package at once - smaller PRs make it easier to review the code
- if you make a PR, reference this issue so we can keep track of the progress
- if you make a PR, label it with `TypeScript` and the package name `wdio-utils` so we can easier find all PRs
- if you make a PR, assign it to the `TypeScript Rewrite` project

Happy rewriting 😅

### CODE REVIEW

File: packages/wdio-sync/webdriverio-core.d.ts
Comment (mgrybyk) [inline packages/wdio-sync/webdriverio-core.d.ts:63]: Please replace function with a proper type

File: packages/wdio-sync/webdriverio-core.d.ts
Comment (mgrybyk) [inline packages/wdio-sync/webdriverio-core.d.ts:66]: Please replace object with Record<string, ...>

File: packages/wdio-utils/src/envDetector.ts
Comment (mgrybyk) [inline packages/wdio-utils/src/envDetector.ts:1]: Should be just webdriver

File: packages/wdio-utils/src/envDetector.ts
Comment (mgrybyk) [inline packages/wdio-utils/src/envDetector.ts:51]: Should be optional according to the code?

**Review by mgrybyk — CHANGES_REQUESTED** [review]

File: packages/wdio-utils/src/envDetector.ts
Comment (christian-bromann) [reply packages/wdio-utils/src/envDetector.ts:1]: I would like to use the "new" types. They will have to be exported properly but I would like to do that once all packages are rewritten. So this is an intermediate solution.

File: packages/wdio-utils/src/initialiseServices.ts
Comment (mgrybyk) [inline packages/wdio-utils/src/initialiseServices.ts:10]: can `object` be replaced with something else, ex `Record<string, unkown>`?

File: packages/wdio-utils/src/test-framework/testInterfaceWrapper.ts
Comment (mgrybyk) [inline packages/wdio-utils/src/test-framework/testInterfaceWrapper.ts:238]: should it be Function instead of any?

File: packages/wdio-utils/src/test-framework/types.ts
Comment (mgrybyk) [inline packages/wdio-utils/src/test-framework/types.ts:27]:
```suggestion
    failedExpectations: Array<Record<string, unkown>>
```

File: packages/wdio-utils/tsconfig.json
Comment (mgrybyk) [inline packages/wdio-utils/tsconfig.json:7]:
```suggestion
    "types": ["webdriverio", "@wdio/logger"]
```

**Review by mgrybyk — CHANGES_REQUESTED** [review]
please apply the changed you made to tsconfig.json to tsconfig.prod.json as well

**Comment by christian-bromann** [comment]
@mgrybyk thanks for the review. I applied all suggestions. Please have another look.

File: packages/wdio-utils/src/envDetector.ts
Comment (mgrybyk) [inline packages/wdio-utils/src/envDetector.ts:14]: Shouls be optional

File: packages/wdio-utils/src/envDetector.ts
Comment (mgrybyk) [inline packages/wdio-utils/src/envDetector.ts:67]:
```suggestion
function isMobile (capabilities?: Capabilities) {
```

File: packages/wdio-utils/src/envDetector.ts
Comment (mgrybyk) [inline packages/wdio-utils/src/envDetector.ts:97]:
```suggestion
function isIOS (capabilities?: DesiredCapabilities) {
```

File: packages/wdio-utils/src/envDetector.ts
Comment (mgrybyk) [inline packages/wdio-utils/src/envDetector.ts:113]:
```suggestion
function isAndroid (capabilities?: Capabilities) {
```

File: packages/wdio-utils/src/envDetector.ts
Comment (mgrybyk) [inline packages/wdio-utils/src/envDetector.ts:130]:
```suggestion
function isSauce (capabilities?: DesiredCapabilities) {
```

File: packages/wdio-utils/src/envDetector.ts
Comment (mgrybyk) [inline packages/wdio-utils/src/envDetector.ts:145]:
```suggestion
function isSeleniumStandalone (capabilities?: DesiredCapabilities) {
```

File: packages/wdio-utils/src/monad.ts
Comment (mgrybyk) [inline packages/wdio-utils/src/monad.ts:11]: Is it possible to deal with `any`?

**Review by mgrybyk — CHANGES_REQUESTED** [review]

File: packages/wdio-utils/src/monad.ts
Comment (christian-bromann) [reply packages/wdio-utils/src/monad.ts:11]: I don't want to import any Puppeteer types just because I had to type it for something that we don't use here:

```js
        /**
         * assign propertiesObject to itself so the client can be recreated
         */
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        const { puppeteer, ...propertiesObjectWithoutPuppeteer } = propertiesObject
        propertiesObject['__propertiesObject__'] = { value: propertiesObjectWithoutPuppeteer }
```

Comment (mgrybyk) [reply packages/wdio-utils/src/monad.ts:11]: Makes sense

File: packages/wdio-utils/src/envDetector.ts
Comment (erwinheitzman) [inline packages/wdio-utils/src/envDetector.ts:54]:
```
        return Boolean(capabilities.chrome || capabilities['goog:chromeOptions'])
```

File: packages/wdio-utils/src/envDetector.ts
Comment (erwinheitzman) [inline packages/wdio-utils/src/envDetector.ts:133]: Is this some kind of bugfix in this PR?

Comment (christian-bromann) [reply packages/wdio-utils/src/envDetector.ts:133]: Yeah, detecting the Sauce background would not have worked using pure WebDriver capabilities with the `webdriver` package

File: packages/wdio-utils/src/shim.ts
Comment (mgrybyk) [inline packages/wdio-utils/src/shim.ts:15]: Is args is any or any array? Not necessarily an object, I guess

File: packages/wdio-utils/tests/__mocks__/@wdio/utils.ts
Comment (mgrybyk) [inline packages/wdio-utils/tests/__mocks__/@wdio/utils.ts:16]: Can any be replaced here?

File: packages/wdio-utils/src/initialiseServices.ts
Comment (erwinheitzman) [inline packages/wdio-utils/src/initialiseServices.ts:51]: log.debug(`initialise custom service "${serviceName.name}"`)

File: packages/wdio-utils/src/initialiseServices.ts
Comment (erwinheitzman) [inline packages/wdio-utils/src/initialiseServices.ts:52]: initialisedServices.push([serviceName as HookFunctions, serviceConfig])

File: packages/wdio-utils/src/initialiseServices.ts
Comment (erwinheitzman) [inline packages/wdio-utils/src/initialiseServices.ts:65]:
```suggestion
        initialisedServices.push([service, serviceConfig, serviceName])
```

File: packages/wdio-utils/src/initialisePlugin.ts
Comment (erwinheitzman) [inline packages/wdio-utils/src/initialisePlugin.ts:10]: add `return X as WebdriverIO.ServiceClass` in the return statements here so you don't need to cast it everywhere else in the code

File: packages/wdio-utils/src/monad.ts
Comment (erwinheitzman) [inline packages/wdio-utils/src/monad.ts:32]:
```suggestion
    const scopeType = SCOPE_TYPES[propertiesObject.scope || 'browser']
```

File: packages/wdio-utils/src/shim.ts
Comment (erwinheitzman) [inline packages/wdio-utils/src/shim.ts:7]: This now changed from null to undefined, is that what we want?

File: packages/wdio-utils/src/test-framework/testFnWrapper.ts
Comment (erwinheitzman) [inline packages/wdio-utils/src/test-framework/testFnWrapper.ts:90]: on line 83, change it to: `let afterArgs = afterFnArgs(this) as [JasmineContext, unknown]` so that you can remove all this castings here

File: packages/wdio-utils/src/test-framework/types.ts
Comment (erwinheitzman) [inline packages/wdio-utils/src/test-framework/types.ts:46]: Let's make a interface where the number is optional for the hooks and specs

File: packages/wdio-utils/src/utils.ts
Comment (erwinheitzman) [inline packages/wdio-utils/src/utils.ts:10]:
```suggestion
export function overwriteElementCommands(propertiesObject: { '__elementOverrides__'?: { value: any }, [key: string]: any }) {
```

File: packages/wdio-utils/src/test-framework/types.ts
Comment (christian-bromann) [reply packages/wdio-utils/src/test-framework/types.ts:46]: what number?

Comment (christian-bromann) [reply packages/wdio-utils/src/test-framework/types.ts:46]: having `[string, Function] |` in line 42 makes this optional

File: packages/wdio-utils/src/shim.ts
Comment (christian-bromann) [reply packages/wdio-utils/src/shim.ts:7]: yes, there is no different in the behavior.

File: packages/wdio-utils/src/utils.ts
Comment (erwinheitzman) [inline packages/wdio-utils/src/utils.ts:217]: Is the 2nd check even needed?

File: packages/wdio-utils/tests/__mocks__/wdio-scoped-service.ts
Comment (erwinheitzman) [inline packages/wdio-utils/tests/__mocks__/wdio-scoped-service.ts:2]:
```suggestion
    isScoped = false
```

Then remove the constructor.
This can be done for similar classes.

File: packages/wdio-utils/tests/envDetector.test.ts
Comment (erwinheitzman) [inline packages/wdio-utils/tests/envDetector.test.ts:14]: double casting?

Comment (erwinheitzman) [reply packages/wdio-utils/tests/envDetector.test.ts:14]: removing the castings doesn't give me any errors, is this needed?

File: packages/webdriver/webdriver.d.ts
Comment (erwinheitzman) [inline packages/webdriver/webdriver.d.ts:184]: Already exists in the types.ts file, like I said in gitter. We should find a standard for ourself to prevent these kind of mistakes :)

File: packages/wdio-utils/src/initialiseServices.ts
Comment (erwinheitzman) [reply packages/wdio-utils/src/initialiseServices.ts:65]: Note, this can only be done if the below comment is resolved first

File: packages/wdio-utils/tests/__mocks__/@wdio/utils.ts
Comment (christian-bromann) [reply packages/wdio-utils/tests/__mocks__/@wdio/utils.ts:16]: not yet, we are waiting for #5909 to land

File: packages/wdio-utils/src/test-framework/testFnWrapper.ts
Comment (christian-bromann) [reply packages/wdio-utils/src/test-framework/testFnWrapper.ts:90]: this if statement is specific to jasmine which is why I have to case it only here

File: packages/wdio-utils/tests/envDetector.test.ts
Comment (christian-bromann) [reply packages/wdio-utils/tests/envDetector.test.ts:14]: it was the only chance for me to case timeout strings correctly, I will remove and see what the build says

File: packages/webdriver/webdriver.d.ts
Comment (christian-bromann) [reply packages/webdriver/webdriver.d.ts:184]: yes, for now we have to maintain two sources of truth and once we have rewritten everything we can clean everything up.

**Comment by christian-bromann** [comment]
🎉

File: packages/wdio-utils/src/initialiseServices.ts
Comment (erwinheitzman) [inline packages/wdio-utils/src/initialiseServices.ts:19]: Add a Service class:

```suggestion
function initialiseServices (services: Service[]): IntialisedService[] {
```

File: packages/wdio-utils/src/initialiseServices.ts
Comment (erwinheitzman) [inline packages/wdio-utils/src/initialiseServices.ts:78]: After creating the Service interface, use it here.

```suggestion
function sanitizeServiceArray (service: WebdriverIO.ServiceEntry): Service[] {
```

File: packages/wdio-utils/tests/utils.test.ts
Comment (mgrybyk) [inline packages/wdio-utils/tests/utils.test.ts:200]: Can some of `@ts-ignore` in this files be removed?

Comment (christian-bromann) [reply packages/wdio-utils/tests/utils.test.ts:200]: the types expect a string but we are testing here if the user passes in nothing how it will react so we have to ignore the types for these

File: packages/wdio-utils/src/utils.ts
Comment (erwinheitzman) [inline packages/wdio-utils/src/utils.ts:165]: The following should work as well.

```suggestion
        if (!require?.main?.paths.includes(localNodeModules)) {
```

**Review by erwinheitzman — APPROVED** [review]
Awesome work! LGTM after @mgrybyk has given his approval as well =]

File: packages/wdio-utils/tests/utils.test.ts
Comment (mgrybyk) [inline packages/wdio-utils/tests/utils.test.ts:214]: is it possible to do something like this `(fs.accessSync as jest.Mock)`?

**Review by mgrybyk — APPROVED** [review]

File: packages/wdio-utils/tests/utils.test.ts
Comment (christian-bromann) [reply packages/wdio-utils/tests/utils.test.ts:214]: Nope, unfortunately not which is why I added it there like that. Happy to revisit all `@ts-config` and refine them once we moved over.

Comment (erwinheitzman) [reply packages/wdio-utils/tests/utils.test.ts:214]: Yes, you can do `jest.Mock<any, any>` if it allows that where the first any is what it put in (believe) and the 2nd any is what is returned if I am correct

**Pull request merged**
